src/metric/rank_metric.cc

```diff
@@ -393,140 +380,230 @@ struct EvalMAP : public EvalRankList {
   }
 };
 
-/*! \brief Cox: Partial likelihood of the Cox proportional hazards model */
-struct EvalCox : public Metric {
- public:
-  EvalCox() = default;
+/*! \brief Area Under Curve, for both classification and rank computed on CPU */
+struct EvalAuc : public Metric {
+ private:
+  // This is used to compute the AUC metrics on the GPU - for ranking tasks and
+  // for training jobs that run on the GPU.
+  std::unique_ptr<xgboost::Metric> auc_gpu_;
+  bool check_gpu_metric_{true};
+
+  template <typename WeightPolicy>
   bst_float Eval(const HostDeviceVector<bst_float> &preds,
                  const MetaInfo &info,
-                 bool distributed) override {
-    CHECK(!distributed) << "Cox metric does not support distributed evaluation";
-    using namespace std;  // NOLINT(*)
+                 bool distributed,
+                 const std::vector<unsigned> &gptr) {
+    const auto ngroups = static_cast<bst_omp_uint>(gptr.size() - 1);
+    // sum of all AUC's across all query groups
+    double sum_auc = 0.0;
+    int auc_error = 0;
+    const auto& labels = info.labels_.ConstHostVector();
+    const std::vector<bst_float>& h_preds = preds.ConstHostVector();
 
-    const auto ndata = static_cast<bst_omp_uint>(info.labels_.Size());
-    const std::vector<size_t> &label_order = info.LabelAbsSort();
+    #pragma omp parallel reduction(+:sum_auc, auc_error) if (ngroups > 1)
+    {
+      // Each thread works on a distinct group and sorts the predictions in that group
+      PredIndPairContainer rec;
+      #pragma omp for schedule(static)
+      for (bst_omp_uint group_id = 0; group_id < ngroups; ++group_id) {
+        // Same thread can work on multiple groups one after another; hence, resize
+        // the predictions array based on the current group
+        rec.resize(gptr[group_id + 1] - gptr[group_id]);
+        #pragma omp parallel for schedule(static) if (!omp_in_parallel())
+        for (bst_omp_uint j = gptr[group_id]; j < gptr[group_id + 1]; ++j) {
+          rec[j - gptr[group_id]] = {h_preds[j], j};
+        }
 
-    // pre-compute a sum for the denominator
-    double exp_p_sum = 0;  // we use double because we might need the precision with large datasets
+        if (omp_in_parallel()) {
+          std::stable_sort(rec.begin(), rec.end(), common::CmpFirst);
+        } else {
+          XGBOOST_PARALLEL_SORT(rec.begin(), rec.end(), common::CmpFirst);
+        }
 
-    const std::vector<bst_float>& h_preds = preds.HostVector();
-    for (omp_ulong i = 0; i < ndata; ++i) {
-      exp_p_sum += h_preds[i];
+        // calculate AUC
+        double sum_pospair = 0.0;
```

Comment:
There is a bit much for me to review in one go. In particular when stuff gets moved its harder to see where the algorithm got changed in the diff. Can we try to separate this into a few PRs, maybe one for moving things around, one for logic for launching/configuring GPU metrics, one per changed metric. Given that a AUC is such a key metric I really want to digest whats going on here.